daemon/cluster/cluster.go

```diff
@@ -1070,23 +1070,40 @@ func (c *Cluster) ServiceLogs(ctx context.Context, input string, config *backend
 		return c.errNoManager(state)
 	}
 
+	var stream swarmapi.Logs_SubscribeLogsClient
 	service, err := getService(ctx, state.controlClient, input)
 	if err != nil {
-		c.mu.RUnlock()
-		return err
-	}
-
-	stream, err := state.logsClient.SubscribeLogs(ctx, &swarmapi.SubscribeLogsRequest{
-		Selector: &swarmapi.LogSelector{
-			ServiceIDs: []string{service.ID},
-		},
-		Options: &swarmapi.LogSubscriptionOptions{
-			Follow: config.Follow,
-		},
-	})
-	if err != nil {
-		c.mu.RUnlock()
-		return err
+		task, err := getTask(ctx, state.controlClient, input)
+		if err != nil {
+			c.mu.RUnlock()
+			logrus.Debug(err)
+			return fmt.Errorf("neither a service nor a task: %s", input)
+		}
+		stream, err = state.logsClient.SubscribeLogs(ctx, &swarmapi.SubscribeLogsRequest{
+			Selector: &swarmapi.LogSelector{
+				TaskIDs: []string{task.ID},
+			},
+			Options: &swarmapi.LogSubscriptionOptions{
+				Follow: config.Follow,
+			},
+		})
+		if err != nil {
+			c.mu.RUnlock()
+			return err
+		}
+	} else {
+		stream, err = state.logsClient.SubscribeLogs(ctx, &swarmapi.SubscribeLogsRequest{
```

Comment:
This code (`state.logsClient.SubscribeLogs`) and the error-handling below is repeated in both "if" and "else"; the only difference is the "Selector" that's used; perhaps just set the selector here, and move the other code outside the if/else.